rfcs/0009-output-commitments/0009-output-commitments.md

```diff
@@ -0,0 +1,107 @@
+
+Suppose we have 2 transactions, the first transaction have 4 outputs and second have 3 outputs, the MMR would be:
+
+<img src="images/01.jpg" width="50%" height="50%">
+
+And we add a transaction with 2 outputs, the MMR would be:
+
+<img src="images/02.jpg" width="80%" height="80%">
+
+Then add a transaction with 4 outputs, the MMR would be:
+
+![](images/03.jpg "MMR with 4 transactions")
+
+Then add a transaction with 0 outputs, the MMR would be:
+
+![](images/04.jpg "MMR with 5 transactions")
+
+Specially, the root of an empty MMR is `H256::zero`.
+
+### Spending
+
+If we want spend an output, we just mark the corresponding CBMT node as `H256::zero` and update the CMBT and MMR.
+
+### Pruning
+
+The MMR can be pruned if there are more than **n** transactions(all the outputs in one transaction is combined into a leaf node) in it. The old transaction is always pruned first. In order to update MMR correctly, if a node is still in the MMR, its sibling should not be pruned.
+
+If **n=3**, the MMR above would be:
+
+![](images/05.jpg "Pruned MMR with 5 transactions")
+
+### Delayed Commitments
+
+A meta is added into the block header called outputs_root, it is the root of MMR.
+
+When a fork happen, it may cost a lot of work on MMR's rollback. In order to avoid this, we delay the commitments **m** blocks, which means the block **i** commits to the TXO set state as of block **i-m**, in other words the outputs_root in block **i** is the root of MMR up until block **i-m**. So the STXOs and UTXOs of latest **m** blocks should be cached.
+
+### Proof
+
+For the sake of illustration, we define **Breadth Order** is that from ***top to bottom*** and ***left to right***.
+
+If an output is pruned, the miner should attach a proof on the block to prove that this output is still unspent. The structure of proof is:
+
+```
+table MMRProof {
+    output_proofs: [CBMTProof],
+    nodes: [H256],
+    left_peaks: [H256],
+}
+```
+
+`CBMTProof` is defined in RFC#0006, it is used for proving that certain outputs inner a transaction is still unspent. The proofs in `output_proofs` are in **descending breadth order** by the node it proved for. The `nodes` is a list of siblings of the nodes along the path that form leaves to peaks, excluding the nodes already in the path, they are also in **descending breadth order**. The `left_peaks` is a list of peaks that can't be calculated directly. The right most peaks should be combined. Also, the peaks are in **descending breadth order**.
+
+If we want to prove the gray nodes in below figure are unspent, the `nodes` in MMRProof would be `[Node4, Node1]` and the `left_peaks` would be `[Node8]`.
+
+![](images/06.jpg "MMR with 5 transactions")
+
+The proofs are attached to the block body with outputs, this can be done by miner or node server. The structure is:
+
+```
+table OutputWithIndex {
+    output: OutPut,
```

Comment:
Should not be `OutPut`. And what is an Output? If it is the entire transaction output, explain it, use the name in current schema file, and link to the latest schema file.